contracts/core/Registry.sol

```diff
@@ -1,47 +1,315 @@
+    enum RoleType {
+        OWNER,
+        MEMBER
+    }
+
+    /// @notice Struct to hold details of an identity
+    struct Identity {
+        bytes32 id;
+        uint nonce;
         string name;
-        string metadata; // @todo update to MetaPtr
-        address attestationAddr;
+        Metadata metadata;
+        address anchor;
+    }
+
+    /// ==========================
+    /// === Storage Variables ====
+    /// ==========================
+
+    /// @notice Identity.id -> Identity
+    mapping(bytes32 => Identity) public identitiesById;
+
+    /// @notice anchor -> Identity.id
+    mapping(address => bytes32) public anchorToIdentityId;
+
+    /// ======================
+    /// ======= Events =======
+    /// ======================
+
+    event IdentityCreated(
+        bytes32 indexed identityId,
+        uint nonce,
+        string name,
+        Metadata metadata,
+        address anchor
+    );
+    event IdentityNameUpdated(
+        bytes32 indexed identityId,
+        string name,
+        address anchor
+    );
+    event IdentityMetadataUpdated(
+        bytes32 indexed identityId,
+        Metadata metadata
+    );
+
+    /// ======================
+    /// ==== Constructor =====
+    /// ======================
+    constructor() {
+        // Empty constructor
+    }
+
+    /// ====================================
+    /// ==== External/Public Functions =====
+    /// ====================================
+
+    /// @notice Retrieve identity by identityId
+    /// @param identityId The identityId of the identity
+    function getIdentity(bytes32 identityId) public view returns (Identity memory) {
+        return identitiesById[identityId];
+    }
+
+    /// @notice Retrieve identity by anchor
+    /// @param anchor The anchor of the identity
+    function getIdentityByAnchor(address anchor) public view returns (Identity memory) {
+        bytes32 identityId = anchorToIdentityId[anchor];
+        return identitiesById[identityId];
+    }
+
+    /// @notice Creates a new identity
+    /// @dev This will also set the attestation address generated from msg.sender and name
+    /// @param _nonce Index of the identity
+    /// @param _name The name of the identity
+    /// @param _metadata The metadata of the identity
+    /// @param _owner The owner of the identity
+    /// @param _members The members of the identity
+    function createIdentity(
+        uint _nonce,
+        string memory _name,
+        Metadata memory _metadata,
+        address _owner,
+        address[] memory _members
+    ) external returns (bytes32) {
+        bytes32 identityId = _generateIdentityId(_nonce);
+
+        if (identitiesById[identityId].id != bytes32(0)) {
+            revert INDEX_NOT_AVAILABLE();
+        }
+
+        Identity memory identity = Identity(
+            identityId,
+            _nonce,
+            _name,
+            _metadata,
+            _generateAnchor(identityId, _name)
+        );
```

Comment:
My paranoia but I'd recommend using named assignment here, like 
```
Identity({
    id: identityId,
    ....
});
```
Seen too many projects change the order of variables in the struct later (say, for gas optimizations) and end up screwing themselves.